pkg/manifest/installer.go

```diff
@@ -390,6 +407,166 @@ func waitForCRDs(objects object.K8sObjects, dryRun bool) error {
+				if err != nil {
+					return false, err
+				}
+				list, err := getPods(cs, ds.Namespace, ds.Spec.Selector.MatchLabels)
+				if err != nil {
+					return false, err
+				}
+				pods = append(pods, list...)
+			case "StatefulSet":
+				sts, err := cs.AppsV1().StatefulSets(o.Namespace).Get(o.Name, metav1.GetOptions{})
+				if err != nil {
+					return false, err
+				}
+				list, err := getPods(cs, sts.Namespace, sts.Spec.Selector.MatchLabels)
+				if err != nil {
+					return false, err
+				}
+				pods = append(pods, list...)
+			case "ReplicaSet":
+				rs, err := cs.AppsV1().ReplicaSets(o.Namespace).Get(o.Name, metav1.GetOptions{})
+				if err != nil {
+					return false, err
+				}
+				list, err := getPods(cs, rs.Namespace, rs.Spec.Selector.MatchLabels)
+				if err != nil {
+					return false, err
+				}
+				pods = append(pods, list...)
+			case "Service":
+				svc, err := cs.CoreV1().Services(o.Namespace).Get(o.Name, metav1.GetOptions{})
+				if err != nil {
+					return false, err
+				}
+				services = append(services, *svc)
+			}
+		}
+		isReady := podsReady(pods) && deploymentsReady(deployments) && servicesReady(services)
+		return isReady, nil
+	})
+
+	if errPoll != nil {
+		logAndPrint("Failed to wait for resources ready: %v", errPoll)
+		return fmt.Errorf("failed to wait for resources ready: %s", errPoll)
+	}
+
+	logAndPrint("Resources are ready.")
+	return nil
+}
+
+func getPods(client kubernetes.Interface, namespace string, selector map[string]string) ([]v1.Pod, error) {
+	list, err := client.CoreV1().Pods(namespace).List(metav1.ListOptions{
+		FieldSelector: fields.Everything().String(),
+		LabelSelector: labels.Set(selector).AsSelector().String(),
+	})
+	return list.Items, err
+}
+
+func podsReady(pods []v1.Pod) bool {
+	for _, pod := range pods {
+		if !isPodReady(&pod) {
+			logAndPrint("Pod is not ready: %s/%s", pod.GetNamespace(), pod.GetName())
+			return false
+		}
+	}
+	return true
+}
+
+func isPodReady(pod *v1.Pod) bool {
+	if len(pod.Status.Conditions) > 0 {
+		for _, condition := range pod.Status.Conditions {
+			if condition.Type == v1.PodReady &&
+				condition.Status == v1.ConditionTrue {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func deploymentsReady(deployments []deployment) bool {
+	for _, v := range deployments {
+		if !(v.replicaSets.Status.ReadyReplicas >= *v.deployment.Spec.Replicas-deploymentutil.MaxUnavailable(*v.deployment)) {
+			logAndPrint("Deployment is not ready: %s/%s", v.deployment.GetNamespace(), v.deployment.GetName())
+			return false
+		}
+	}
+	return true
+}
+
+func servicesReady(svc []v1.Service) bool {
+	for _, s := range svc {
+		if s.Spec.Type == v1.ServiceTypeExternalName {
+			continue
+		}
+		if s.Spec.ClusterIP != v1.ClusterIPNone && s.Spec.ClusterIP == "" {
+			logAndPrint("Service is not ready: %s/%s", s.GetNamespace(), s.GetName())
+			return false
+		}
+		if s.Spec.Type == v1.ServiceTypeLoadBalancer && s.Status.LoadBalancer.Ingress == nil {
```

Comment:
@ostromart One thing I'm not sure is whether we should check `loadBalancerIP` is available or not for loadBalancer sevice(istio-ingressgateway), because many clusters don't provide loadBalancer implementation, resulting in `pending` in the `loadBalancerIP`; I think end users don't want to wait because of `loadBalancerIP`, what do you think?